tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +395,960 @@ impl ServerKey {
+        } else {
+            self.advanced_add_assign_with_carry_sequential_parallelized(
+                lhs,
+                rhs,
+                input_carry,
+                requested_flag,
+            )
+        }
+    }
+
+    /// Computes the result of `lhs += rhs + input_carry`
     ///
-    /// - The parameters have 4 bits in total
-    /// - Adding rhs to lhs must not consume more than one carry
+    /// This uses the sequential algorithm to propagate the carries
     ///
-    /// # Output
+    /// - `lhs` and `rhs` must have the same `len()`, empty is allowed
+    /// - `blocks of lhs` and `rhs` must all be without carry
+    /// - blocks must have at least one bit of message and one bit of carry
     ///
-    /// - lhs will have its carries empty
-    pub(crate) fn unchecked_add_assign_parallelized_low_latency<T>(
+    /// Returns `Some(...)` if requested_flag != ComputationFlags::None
+    pub(crate) fn advanced_add_assign_with_carry_sequential_parallelized(
         &self,
-        lhs: &mut T,
-        rhs: &T,
-    ) -> Ciphertext
-    where
-        T: IntegerRadixCiphertext,
-    {
-        let degree_after_add_does_not_go_beyond_first_carry = lhs
-            .blocks()
-            .iter()
-            .zip(rhs.blocks().iter())
-            .all(|(bl, br)| {
-                let degree_after_add = bl.degree.get() + br.degree.get();
-                degree_after_add < (self.key.message_modulus.0 * 2)
+        lhs: &mut [Ciphertext],
+        rhs: &[Ciphertext],
+        input_carry: Option<&BooleanBlock>,
+        requested_flag: OutputFlag,
+    ) -> Option<BooleanBlock> {
+        assert_eq!(
+            lhs.len(),
+            rhs.len(),
+            "Both operands must have the same number of blocks"
+        );
+
+        if lhs.is_empty() {
+            return if requested_flag == OutputFlag::None {
+                None
+            } else {
+                Some(self.create_trivial_boolean_block(false))
+            };
+        }
+
+        let mut carry = input_carry.map_or_else(
+            || self.key.create_trivial(0),
+            |boolean_block| boolean_block.0.clone(),
+        );
+
+        // 2_2, 3_3, 4_4
+        // If we have at least 2 bits and at least as much carries
+        //
+        // The num blocks == 1 + requested_flag == OverflowFlag will actually result in one more
+        // PBS of latency than num_blocks == 1 && requested_flag != OverflowFlag
+        //
+        // It happens because the computation of the overflow flag requires 2 steps,
+        // and we insert these two steps in parallel to normal carry propagation.
+        // The first step is done when processing the first block,
+        // the second step is done when processing the last block.
+        // So if the number of block is smaller than 2 then,
+        // the overflow computation adds additional layer of PBS.
+        if self.key.message_modulus.0 >= 4 && self.key.carry_modulus.0 >= self.key.message_modulus.0
```

Comment:
should the first block of the if be a function and the else if another function ?